proposals/010-workflow-migrations.md

```diff
@@ -0,0 +1,303 @@
+    // if not provided then possible errors will be thrown
+    map<string, VariableValue> migration_variables = 3;
+}
+```
+
+
+
+For every thread we are migrating we need to know its exact destination in the new `wfSpec`so every thread migration has a new_thread_name -- name of the thread in the new `wfSpec`.
+
+Additionally we need to know what node within the new thread we are migrating to and for that reason we have node_migrations.
+
+migration_variables allows users to inject variableValues into the new Spec. This usefull when variableValues would have been derived prior to the migration node 
+
+```
+//  How to migrate Node 
+message NodeMigrationPlan {
+
+    // Currently just migrate to node with this name 
+    string new_node = 1;
+}
+```
+As of now just the new_node name will be sufficient. A check will be done making sure nodes are of same sub type before migrating.
+
+### Loose `WfRunMigrationPlan` Criteria
+
+This implementation provides loose enforcement of rules when registering the `wfRunMigrationPlan` metadata object, since at registration time we do not yet know the runtime state of any `wfRun` that will be migrated or the version it will migrate to.
+
+This design allows users to define a single `wfRunMigrationPlan` object and apply that migration to multiple `wfRuns` without having to redefine the migration plan for each migration.
+
+### Additional Proto changes
+
+```
+enum GetableClassEnum {
+    WF_RUN_MIGRATION_PLAN = 21;
+}
+```
+
+Getting Object:
+
+```
+  // Get Migration Plan
+  rpc getMigrationPlan(MigrationPlanId) returns (WfRunMigrationPlan) {}
+```
+
+```
+message MigrationPlanId {
+    string name = 1;
+}
+```
+
+## Migrating a `wfRun`
+
+Now that the `wfRunMigrationPlan` has been registered, we can migrate a `wfRun` using the following RPC and schema implementation.
+
+```
+rpc MigrateWfRun(MigrateWfRunRequest) returns (WfRunId) {}
+```
+
+```
+message MigrateWfRunRequest {
+  // Migration Plan to use
+  MigrationPlanId migration_plan_id = 1; 
+
+  // WfRun to Migrate
+  WfRunId wf_run_id = 2;
+
+  // What version to migrate to
+  int32 revision_number = 3;
+  int32 major_version_number = 4;
+
+}
+```
+To perform a migration, we need the name of the migration plan, the `wfRunId` of the `wfRun` being migrated, the revision number, and the major_version_number of the new specification to which it will migrate.
+
+The migration schema itself is fairly straightforward. The complexity arises in the server-side logic executed during a migration.
+
+### Option 2:
+
+In the first impl you must provide the migration variables for every thread in the `wfRunMigrationPlan`.
+Another approach to enable more flexibility would be to define the variables in the `migrateWfRunRequest`.
+
+### Proto
+
+We define the set of variables that are derived before the migration node but needed after the migration node
+
+```
+// How to migrate a thread 
+message ThreadMigrationPlan {
```

Comment:
Same comment here. Are we dealing with `ThreadSpec` or `ThreadRun`? To be obnoxious, I might say

> This is a plan to migrate a `ThreadRun` from one `ThreadSpec` to another `ThreadSpec` in a newer `WfSpec`.